ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/PostCommitValidation.scala

```diff
@@ -0,0 +1,230 @@
+
+      // Collect all the identifiers of contracts that have not been divulged in the current transaction
+      val maximumLedgerEffectiveTime =
+        data.lookupMaximumLedgerTime(
+          transaction.fold(Set.empty[ContractId]) {
+            case (ids, (_, e: Exercise)) if !divulged(e.targetCoid) => ids + e.targetCoid
+            case (ids, (_, f: Fetch)) if !divulged(f.coid) => ids + f.coid
+            case (ids, (_, l: LookupByKey)) => l.result.filterNot(divulged).fold(ids)(ids + _)
+            case (ids, _) => ids
+          }
+        )
+
+      // Query the committed contracts store for the maximum ledger effective time of the
+      // collected contracts and returns an error if that is strictly greater than the
+      // ledger effective time of the current transaction.
+      maximumLedgerEffectiveTime
+        .map(
+          _.filter(_.isAfter(transactionLedgerEffectiveTime)).fold(Set.empty[RejectionReason])(
+            contractLedgerEffectiveTime => {
+              Set(
+                CausalMonotonicityViolation(
+                  contractLedgerEffectiveTime = contractLedgerEffectiveTime,
+                  transactionLedgerEffectiveTime = transactionLedgerEffectiveTime,
+                )
+              )
+            }
+          )
+        )
+        .getOrElse(Set(UnknownContract))
+    }
+
+    private def validateKeyUsages(transaction: Transaction, submitter: Party)(
+        implicit connection: Connection): Set[RejectionReason] = {
+      transaction
+        .fold(State.empty(data, submitter)) {
+          case (state, (_, node)) => validateKeyUsages(node, state)
+        }
+        .errors
+    }
+
+    private def validateKeyUsages(
+        node: Node,
+        state: State,
+    )(implicit connection: Connection): State = {
+
+      node match {
+        case c: Create =>
+          state.validateCreate(c.key.map(convert(c.coinst.template, _)), c.coid)
+        case l: LookupByKey =>
+          state.validateLookupByKey(convert(l.templateId, l.key), l.result)
+        case e: Exercise if e.consuming =>
+          state.removeKeyIfDefined(e.key.map(convert(e.templateId, _)))
+        case _ =>
+          // fetch and non-consuming exercise nodes don't need to validate
+          // anything with regards to contract keys and do not alter the
+          // state in a way which is relevant for the validation of
+          // subsequent nodes
+          state
+      }
+    }
+  }
+
+  /**
+    * Represents the state of an ongoing validation.
+    * It must be carried over as the transaction is
+    * validated one node at a time in pre-order
+    * traversal for this to make sense.
+    *
+    * @param errors Accumulates errors retrieved so far while validating
+    * @param contracts All contracts created as part of the current transaction
+    * @param removed Ensures indexed transactions are not referred to if they are removed in the current transaction
```

Comment:
Should `indexed transactions` rather be `indexed contracts` or `indexed keys`?

Reply:
Absolutely, thanks!

Reply:
Addressed by https://github.com/digital-asset/daml/pull/5737/commits/83d123331c49e94ed8209fd263cde54edac3957a